02-peer-protocol.md

```diff
@@ -49,11 +68,468 @@ transaction is confirmed are also not persistent - until you know the script
+
+#### Requirements
+
+The sending node:
+  - MUST NOT send a `tx_remove` with a `serial_id` it did not add
+    to the transaction or has already removed
+
+The receiving node:
+  - MUST remove the indicated input or output from the transaction
+  - MUST fail the negotiation if:
+    - the input or output identified by the `serial_id` was not added by the
+      sender
+    - the `serial_id` does not correspond to a currently added input (or output)
+
+### The `tx_complete` Message
+
+This message signals the conclusion of a peer's transaction
+contributions.
+
+1. type: 70 (`tx_complete`)
+2. data:
+    * [`channel_id`:`channel_id`]
+
+#### Requirements
+
+The nodes:
+  - MUST send this message in succession to conclude this protocol
+
+The receiving node:
+  - MUST use the negotiated inputs and outputs to construct a transaction
+  - MUST fail the negotiation if:
+    - the peer's total input satoshis is less than their outputs
+    - the peer's paid feerate does not meet or exceed the agreed `feerate`,
+      (based on the `minimum fee`).
+    - if is the *non-initiator*:
+      - the *initiator*'s fees do not cover the `common` fields
+    - there are more than 252 inputs
+    - there are more than 252 outputs
+    - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)
+
+#### Rationale
+
+To signal the conclusion of exchange of transaction inputs and outputs.
+
+Upon successful exchange of `tx_complete` messages, both nodes
+should construct the transaction and proceed to the next portion of the
+protocol. For channel establishment v2, exchanging commitment transactions.
+
+For the `minimum fee` calculation see [BOLT #3](03-transactions.md#calculating-fees-for-collaborative-transaction-construction).
+
+The maximum inputs and outputs are capped at 252. This effectively fixes
+the byte size of the input and output counts on the transaction to one (1).
+
+### The `tx_signatures` Message
+
+1. type: 71 (`tx_signatures`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`sha256`:`txid`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness_stack`:`witness_stack`]
+
+1. subtype: `witness_stack`
+2. data:
+    * [`u16`:`num_input_witness`]
+    * [`num_input_witness*witness_element`:`witness_element`]
+
+1. subtype: `witness_element`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness`]
+
+#### Requirements
+
+The sending node:
+  - if it has the lowest total satoshis contributed, as defined
+    by total `tx_add_input` values:
+    - MUST transmit their `tx_signatures` before the peer
+  - MUST order the `witness_stack`s by the `serial_id` of the input they
+    correspond to
+  - number of `witness_stack`s MUST equal the number of inputs they added
+
+The receiving node:
+  - MUST fail the negotiation if:
+    - if they have not sent a `tx_signatures` message and they are the
+      peer with the lowest total satoshis contributed, as defined by
+      `tx_add_input` values
```

Comment:
Right, I fully agree, there is no reason to fail if we receive our counterparty's signatures early, this requirement should be removed. Eclair already supports receiving early `tx_signatures` (since as you mention, it's clearly a good thing for our node to receive the remote sigs).